File: services/server/src/safe_socket/safe_socket.py
import socket


def recv_all(sock: socket.socket, size: int) -> bytes:
    """Receive exactly size bytes, retrying successful partial operations.

    Socket exceptions are propagated. An empty result before the requested size
    is complete indicates that the peer closed the connection prematurely.
    """

    if size < 0:
        raise ValueError(f"invalid read size {size}")

    data = bytearray()

    while len(data) < size:
        recv_size = size - len(data)
        # Socket errors are deliberately propagated. Only successful partial
        # receives are retried; retrying an arbitrary error could loop forever.
        chunk = sock.recv(recv_size)
        if chunk == b"":
            raise EOFError(f"unexpected EOF after {len(data)} of {size} bytes received")

        data.extend(chunk)

    return bytes(data)


def send_all(sock: socket.socket, data: bytes) -> None:
    """Send every byte in data, retrying successful partial operations.

    Socket exceptions are propagated, while a zero-byte result without an
    exception is retried until the requested transfer is complete.
    """

    view = memoryview(data)
    total_sent = 0

    while total_sent < len(view):
        # Socket errors are deliberately propagated. Only successful partial
        # sends are retried; retrying an arbitrary error could loop forever.
        sent = sock.send(view[total_sent:])

        total_sent += sent

    return None

File: services/client/src/safe_socket/safe_socket.go
package safe_socket

import (
	"fmt"
	"io"
)

// SendAll keeps writing the remaining suffix until every byte is transferred.
// Writer errors are returned immediately; successful zero-byte writes are retried.
func SendAll(socket io.Writer, data []byte) error {
	totalSent := 0

	for totalSent < len(data) {
		sent, err := socket.Write(data[totalSent:])
		totalSent += sent
		// Retry only successful short writes. An arbitrary error may be
		// permanent, so retrying it could hide the failure or loop forever.
		if err != nil {
			return err
		}

	}

	return nil
}

// RecvAll reads exactly size bytes, preserving bytes returned together with an
// error and reporting an early EOF as io.ErrUnexpectedEOF.
func RecvAll(socket io.Reader, size int) ([]byte, error) {
	if size < 0 {
		return nil, fmt.Errorf("invalid read size %d", size)
	}

	buff := make([]byte, size)
	totalReceived := 0

	for totalReceived < size {
		received, err := socket.Read(buff[totalReceived:])
		totalReceived += received
		if totalReceived == size {
			return buff, nil
		}
		// Keep bytes returned together with an error, but propagate that error
		// if the requested message is still incomplete. Its cause may be permanent.
		if err != nil {
			if err == io.EOF {
				return nil, io.ErrUnexpectedEOF
			}
			return nil, err
		}

	}

	return buff, nil
}
